parachain/pallets/outbound-queue/src/lib.rs

```diff
@@ -314,94 +397,56 @@ pub mod pallet {
 			T::GasMeter::MAXIMUM_BASE_GAS + T::GasMeter::maximum_required(command)
 		}
 
-		/// Fee in DOT for delivering a message.
-		pub(crate) fn delivery_fee(command: &Command) -> T::Balance {
-			let max_gas_used = Self::maximum_overall_required_gas(command);
-			T::DeliveryFeePerGas::get().saturating_mul(max_gas_used.into())
-		}
-
-		/// Maximum refund in Ether for delivering a message
-		pub(crate) fn maximum_refund(command: &Command) -> u128 {
-			let max_gas_used = Self::maximum_overall_required_gas(command);
-			T::DeliveryRefundPerGas::get().saturating_mul(max_gas_used as u128)
-		}
-	}
-
-	impl<T: Config> OutboundQueueTrait for Pallet<T> {
-		type Ticket = OutboundQueueTicket<MaxEnqueuedMessageSizeOf<T>>;
-		type Balance = T::Balance;
-
-		fn validate(message: &Message) -> Result<(Self::Ticket, Fees<Self::Balance>), SendError> {
-			// The inner payload should not be too large
-			let payload = message.command.abi_encode();
-
-			// Create a message id for tracking progress in submission pipeline
-			let message_id: MessageHash = sp_io::hashing::blake2_256(&(message.encode())).into();
-
-			ensure!(
-				payload.len() < T::MaxMessagePayloadSize::get() as usize,
-				SendError::MessageTooLarge
+		/// Calculate fee in native currency for delivering a message.
+		pub(crate) fn calculate_fee(command: &Command) -> Fee<T::Balance> {
+			let max_gas = Self::maximum_overall_required_gas(command);
+			let remote = Self::calculate_remote_fee(
+				max_gas,
+				Self::fee_config().fee_per_gas,
+				Self::fee_config().reward,
 			);
 
-			let base_fee = T::WeightToFee::weight_to_fee(
-				&T::WeightInfo::do_process_message()
-					.saturating_add(T::WeightInfo::commit_one_message()),
-			);
-			let delivery_fee = Self::delivery_fee(&message.command);
-
-			let command = message.command.clone();
-			let fee = Fees { base: base_fee, delivery: delivery_fee };
+			let remote = FixedU128::from(remote)
+				.checked_div(&Self::fee_config().exchange_rate)
+				.expect("exchange rate is not zero; qed")
+				.into_inner()
+				.checked_div(FixedU128::accuracy())
+				.expect("accuracy is not zero; qed");
```

Comment:
Use `FixedU128::from_inner(remote)` then the last step `checked_div(FixedU128::accuracy())` can be ignored.

Reply:
While merging the exchange rate and decimal conversion into a single factor may simplify the code, it increases complexity for governance, who now need to calculate and review the math manually. This complexity increases the chance of errors.